src/libdredd/src/mutation_replace_expr.cc

```diff
@@ -0,0 +1,314 @@
+// Copyright 2022 The Dredd Project Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "libdredd/mutation_replace_expr.h"
+
+#include <cassert>
+#include <sstream>
+
+#include "clang/AST/ASTContext.h"
+#include "clang/AST/ASTTypeTraits.h"
+#include "clang/AST/Expr.h"
+#include "clang/AST/ParentMapContext.h"
+#include "clang/AST/Type.h"
+#include "clang/Basic/LangOptions.h"
+#include "clang/Basic/SourceLocation.h"
+#include "clang/Lex/Preprocessor.h"
+#include "clang/Rewrite/Core/Rewriter.h"
+#include "libdredd/util.h"
+#include "llvm/ADT/StringRef.h"
+
+namespace dredd {
+dredd::MutationReplaceExpr::MutationReplaceExpr(const clang::Expr& expr)
+    : expr_(expr) {}
+
+std::string MutationReplaceExpr::GetFunctionName(
+    clang::ASTContext& ast_context) const {
+  std::string result = "__dredd_replace_expr_";
+
+  if (expr_.isLValue()) {
+    clang::QualType qualified_type = expr_.getType();
+    if (qualified_type.isVolatileQualified()) {
+      assert(expr_.getType().isVolatileQualified() &&
+             "Expected expression to be volatile-qualified since subexpression "
+             "is.");
+      result += "volatile_";
+    }
+  }
+
+  // A string corresponding to the expression forms part of the name of the
+  // mutation function, to differentiate mutation functions for different
+  // types
+  result += SpaceToUnderscore(expr_.getType()
+                                  ->getAs<clang::BuiltinType>()
+                                  ->getName(ast_context.getPrintingPolicy())
+                                  .str());
+
+  if (expr_.isLValue()) {
+    result += "_lvalue";
+  }
+
+  return result;
+}
+
+void MutationReplaceExpr::GenerateUnaryOperatorInsertion(
+    const std::string& arg_evaluated, const clang::Expr& expr_,
+    const clang::BuiltinType& exprType, std::stringstream& new_function,
+    int& mutant_offset) {
+  if (expr_.isLValue() && !(expr_.getType().isConstQualified() ||
+                            expr_.getType()->isBooleanType())) {
+    // In the boolean case, ++ is redundant since -- is sufficient for flipping
+    // the booleans value.
```

Comment:
@JamesLee-Jones please check whether this is needed and get rid of it if it's not.

Reply:
Done.